server/base/src/main/java/org/apache/accumulo/server/conf/codec/VersionedProperties.java

```diff
@@ -0,0 +1,254 @@
+    this.dataVersion = dataVersion;
+    this.timestamp = timestamp;
+    if (Objects.nonNull(props)) {
+      this.props = new ImmutableMap.Builder<String,String>().putAll(props).build();
+    } else {
+      this.props = new ImmutableMap.Builder<String,String>().build();
+    }
+  }
+
+  /**
+   * Get an unmodifiable map with all property key,value pairs.
+   *
+   * @return An unmodifiable view of the property key, value pairs.
+   */
+  public Map<String,String> getProperties() {
+    return props;
+  }
+
+  /**
+   * Get the current data version. The version should match the node version of the stored data. The
+   * value should be used on data writes as the expected version. If the data write fails do to an
+   * unexpected version, it signals that the node version has changed since the instance was
+   * instantiated and encoded.
+   *
+   * @return 0 for initial version, otherwise the data version when the properties were serialized.
+   */
+  public int getDataVersion() {
+    return Math.max(dataVersion, 0);
+  }
+
+  /**
+   * Calculates the version that should be stored when serialized. The serialized version, when
+   * stored, should match the version that will be assigned. This way, data reading the serialized
+   * version can compare the stored version with the node version at any time to detect if the node
+   * version has been updated.
+   * <p>
+   * The initialization of the data version to a negative value allows this value to be calculated
+   * correctly for the first serialization. On the first store, the expected version will be 0.
+   *
+   * @return the next version number that should be serialized, or 0 if this is the initial version.
+   */
+  public int getNextVersion() {
+    return Math.max(dataVersion + 1, 0);
+  }
+
+  /**
+   * The timestamp of the instance when created or last modified.
+   *
+   * @return the timestamp of the instance.
+   */
+  public Instant getTimestamp() {
+    return Instant.from(timestamp);
+  }
+
+  /**
+   * The timestamp formatted as an ISO 8601 string with format of
+   * {@code YYYY-MM-DDTHH:mm:ss.SSSSSSZ}
+   *
+   * @return a formatted timestamp string.
+   */
+  public String getTimestampISO() {
+    return tsFormatter.format(timestamp);
+  }
+
+  /**
+   * Update a single property. If a property already exists it is overwritten.
+   * <p>
+   * It is much more efficient to add multiple properties at a time rather than one by one. Because
+   * instances of this class are immutable, the creates a new copy of the properties. Other
+   * processes will continue to see original values retrieved from the data store. Other processes
+   * will be updated when the instance is encoded and stored in the data store.
+   *
+   * @param key
+   *          the property name.
+   * @param value
+   *          the property value.
+   * @return A new instance of this class with the property added or updated.
+   */
+  public VersionedProperties update(final String key, final String value) {
```

Comment:
I'm wondering if update() could be renamed to something that implies that the properties are being added. "update" could mean other things too. Not a big deal just thought I would bring it up.

Reply:
Actually not so sure. "update" does make sense when a property is being replaced/updated but not so sure when a new prop is being added.

Reply:
Its an add or an update.  New key, values are added, but if the key exists the value is replaced.  An alternative might be addOrUpdate, or upsert?

Reply:
> An alternative might be addOrUpdate, or upsert?

`addOrUpdate` seems pretty intuitive to me